cli/src/evm_commands.rs

```diff
@@ -0,0 +1,227 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::{
+	get_layer_two_nonce,
+	trusted_command_utils::{get_identifiers, get_pair_from_str},
+	trusted_commands::TrustedArgs,
+	trusted_operation::perform_trusted_operation,
+	Cli,
+};
+use codec::Decode;
+use ita_stf::{
+	evm_helpers::evm_create_address, Index, KeyPair, TrustedCall, TrustedGetter, TrustedOperation,
+};
+use itp_types::AccountId;
+use log::*;
+use pallet_evm::{AddressMapping, HashedAddressMapping};
+use sp_core::{crypto::Ss58Codec, Pair, H160, H256, U256};
+use sp_runtime::traits::BlakeTwo256;
+use std::{string::ToString, vec::Vec};
+use substrate_api_client::utils::FromHexString;
+
+macro_rules! get_layer_two_evm_nonce {
+	($signer_pair:ident, $cli:ident, $trusted_args:ident ) => {{
+		let top: TrustedOperation = TrustedGetter::evm_nonce($signer_pair.public().into())
+			.sign(&KeyPair::Sr25519($signer_pair.clone()))
+			.into();
+		let res = perform_trusted_operation($cli, $trusted_args, &top);
+		let nonce: Index = if let Some(n) = res {
+			if let Ok(nonce) = Index::decode(&mut n.as_slice()) {
+				nonce
+			} else {
+				0
+			}
+		} else {
+			0
+		};
+		debug!("got layer two nonce: {:?}", nonce);
+		nonce
+	}};
+}
+
+#[derive(Subcommand)]
+pub enum EvmCommands {
+	/// Create smart contract
+	EvmCreate {
+		/// Sender's incognito AccountId in ss58check format
+		from: String,
+
+		/// Smart Contract in Hex format
+		smart_contract: String,
+	},
+
+	/// Read smart contract storage
+	EvmRead {
+		/// Sender's incognito AccountId in ss58check format
+		from: String,
+
+		/// Execution address of the smart contract
+		execution_address: String,
+	},
+
+	/// Create smart contract
+	EvmCall {
+		/// Sender's incognito AccountId in ss58check format
+		from: String,
+
+		/// Execution address of the smart contract
+		execution_address: String,
+
+		/// Function hash
+		function: String,
+	},
+}
+
+pub fn match_evm_commands(cli: &Cli, trusted_args: &TrustedArgs, evm_command: &EvmCommands) {
+	match &evm_command {
+		EvmCommands::EvmCreate { from, smart_contract } =>
+			evm_create(cli, trusted_args, from, smart_contract),
+		EvmCommands::EvmRead { from, execution_address } =>
+			evm_read_storage(cli, trusted_args, from, execution_address),
+		EvmCommands::EvmCall { from, execution_address, function } =>
+			evm_call(cli, trusted_args, from, execution_address, function),
+	}
+}
```

Comment:
I think I would like the uniform pattern here that I also applied in the exchange oracle.

```rust

impl EvmSubcommand {
   pub fn run(&self, cli: Cli, trusted_args: TrustedArgs) {
     	match self {
			EvmSubcommand::CreateContract(cmd) => cmd.run(cli, trusted_args),
            ....
		}
   }
}
```

Another question: why do we pass `trusted_args` separately? Shouldn't it be part of the `cli` too?

Reply:
trusted args only exist for the trusted subcommands.